fuel/datasets/hdf5.py

```diff
@@ -64,3 +67,127 @@ def get_data(self, state=None, request=None):
+class H5PYDataset(Dataset):
+    """An h5py-fueled HDF5 dataset.
+
+    This dataset class assumes a particular file layout:
+
+    * Data sources reside in the root group, and their names define the
+      source names.
+    * The dataset is not explicitly split. Instead, splits are defined as
+      attributes of the root group. They're expected to be numpy arrays of
+      shape (2,), with the first element being the starting point
+      (inclusive) of the split and the last element being the stopping
+      point (exclusive) of the split.
+
+    The `which_set`, `start` and `stop` parameters work together in the
+    following way:
+
+    * `which_set` is resolved first. If it is `None`, the whole dataset is
+      used.
+    * `start` and `stop` define a slice *within the context of*
+      `which_set`.
+
+    Parameters
+    ----------
+    path : str
+        Path to the HDF5 file.
+    which_set : str, optional
+        Name of the root group attribute containing the split information.
+        Defaults to `None`, in which case the whole dataset is used.
+    start : int
+        Start index *within the split*.
+    stop : int
+        Stop index *within the split*.
+    driver : str, optional
+        Low-level driver to use. Defaults to `None`. See h5py
+        documentation for a complete list of available options.
+
+    """
+    ref_counts = dict()
+
+    def __init__(self, path, which_set=None, start=None, stop=None,
+                 load_in_memory=False, driver=None, **kwargs):
+        self.path = path
+        self.which_set = which_set
+        self.start = start
+        self.stop = stop
+        self.load_in_memory = load_in_memory
+        self.driver = driver
+        self.load()
+        super(H5PYDataset, self).__init__(**kwargs)
+
+    def _get_file_id(self):
+        try:
+            return next(
+                filter(lambda x: x.name == self.path, self.ref_counts.keys()))
+        except StopIteration:
+            return self.path
+
+    def load(self):
+        handle = self._out_of_memory_open()
+        self.provides_sources = list(handle.keys())
+        shapes = [data_source.shape for data_source in handle.values()]
+        if any(s[0] != shapes[0][0] for s in shapes):
+            raise ValueError("sources have different lenghts")
+        start, stop = (handle.attrs[self.which_set] if self.which_set
+                       else (0, shapes[0][0]))
+        self._start = start if self.start is None else start + self.start
+        self._stop = stop if self.stop is None else start + self.stop
+        self.num_examples = self._stop - self._start
+        self.data_sources = ([data_source[self._start: self._stop] for
+                              data_source in handle.values()]
+                             if self.load_in_memory else None)
+        self._out_of_memory_close(handle)
+
+    def open(self):
+        return (self._in_memory_open() if self.load_in_memory
```

Comment:
Is there a reason for not just returning None here? WHy have in_memory_open and in_memory_close at all?